hammerblade/torch/kernel/kernel_addmm.cpp

```diff
@@ -1,29 +1,183 @@
 //====================================================================
 // addmm kernel
-// 03/09/2020 YOUR NAME HERE Lin Cheng (YOU EMAIL HERE, lc873@cornell.edu)
+// 03/09/2020 Kexin Zheng, Lin Cheng (kz73@cornell.edu, lc873@cornell.edu)
 //====================================================================
 
 #include <kernel_common.hpp>
+#include <cmath>
+#define BLOCK_DIM 10 // sqrt(4KB/4 byte/4 data matrix) = 15 max 
 
 extern "C" {
 
   __attribute__ ((noinline))  int tensorlib_addmm(
-          bsg_tensor_t* self,
-          bsg_tensor_t* mat1,
-          bsg_tensor_t* mat2,
-          float* beta,
-          float* alpha) {
+          bsg_tensor_t* _result,
+          bsg_tensor_t* _self,
+          bsg_tensor_t* _mat1,
+          bsg_tensor_t* _mat2,
+          float* _beta,
+          float* _alpha) {
     // TODO: Convert uint32_t pointers to correct types
+    auto self = BSGTensor<float>(_self);
+    auto mat1 = BSGTensor<float>(_mat1);
+    auto mat2 = BSGTensor<float>(_mat2);
+    auto result = BSGTensor<float>(_result);
+    float beta = *_beta;
+    float alpha = *_alpha;
+
     // Start profiling
     bsg_cuda_print_stat_kernel_start();
+
     // TODO: Implement addmm
+
+    // v2: single tile, use blocking
+    int r1 = mat1.dim(0);
+    int c1 = mat1.dim(1);
+    int r2 = mat2.dim(0);
+    int c2 = mat2.dim(1);
+    assert(c1 == r2);
```

Comment:
`assert` calls `exit` which has undefined behavior on the device. Since this is device code, we'd want to exit using device mechanisms. HammerBlade's exit mechanisms is to call `bsg_fail()`. Lin implemented an assertion wrapper around it called `bsg_assert`. You could use this here...

Reply:
Will do!